### PR TITLE
557 : Implement CSV Import of inventory

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/OpenSRP/opensrp-server-web/issues/557

### CODE REVIEW

File: src/main/java/org/opensrp/service/StockService.java [hunk 1]
```diff
@@ -138,7 +168,12 @@ public void addInventory(Inventory inventory, String userName) {
 			throw new IllegalArgumentException(
 					"Invalid Product Name was selected");
 		}
-		Stock existingStock = findByIdentifierAndServicePointId(String.valueOf(productCatalogue.getUniqueId()),inventory.getServicePointId());
```
Comment (githengi) [inline src/main/java/org/opensrp/service/StockService.java:141]: Why is this removed

**Review by githengi — CHANGES_REQUESTED** [review]
For Donor and Unicef Section. We should remove the enums and Define this in settings

This will it make it easy to define the values without needing code changes.

This is example of payload to register a donor. You can then retrieve the settings and validate as you are doing using enums

```
{
  "settingConfigurations": [
    {
      "identifier": "inventory_donors",
      "settings": [
        {
          "description": "",
          "label": "Natcom Germany",
          "type": "SettingConfiguration",
          "value": "Natcom Belgium",
          "key": "NATCOM_BELGIUM"
        },
        {
          "description": "",
          "label": "Natcom Germany",
          "type": "SettingConfiguration",
          "value": "Natcom Germany",
          "key": "NATCOM_GERMANY"
        }
      ],
      "type": "SettingConfiguration"
    }
  ]
}```

**Review by rehammuzzamil — COMMENTED** [review]
> settingConfigurations

@githengi I have a question regarding this, should I create a script for this? As this would be a one-time activity. Or do we need to use an existing end-point of SettingsResource.java to persist the donor and UNICEF sections?

File: src/main/java/org/opensrp/service/StockService.java [hunk 2]
```diff
@@ -185,6 +218,108 @@ public void deleteStock(Long id) {
...
+					rowCount++;
+					inventory = createInventoryObject(csvdata);
+					addInventory(inventory, userName);
+					rowsProcessed++;
+				}
+				catch (Exception e) {
+					failedRecordSummary = new FailedRecordSummary();
+					List<String> validationError = new ArrayList<>();
+					failedRecordSummary.setRowNumber(rowCount);
+					validationError.add("Exception occurred while converting and persisting of Stock data : " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/StockService.java:249]: Noted

File: src/main/java/org/opensrp/service/StockService.java [hunk 3]
```diff
@@ -185,6 +218,108 @@ public void deleteStock(Long id) {
...
+		FailedRecordSummary failedRecordSummary;
+		List<FailedRecordSummary> failedRecordSummaries = new ArrayList<>();
+		Inventory inventory;
+		Integer totalRows = csvStocks.size();
+		Integer rowsProcessed = 0;
+
+		try {
+			failedRecordSummaries = validateInventoryData(csvStocks);
+		}
+		catch (ParseException e) {
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/StockService.java:233]: I will add the error in the `failedRecordSummaries` list in the catch block as done on Line no: 251 
In this way, it would ensure if there is an error or exception occurred during validation!
@githengi

[the remark this answers is not in the document]
Comment (githengi) [reply src/main/java/org/opensrp/service/StockService.java:233]: noted

**Comment by githengi** [comment]
> question regarding this, should I create a script for this? As this would be a one-time activity. Or do we need to use an existing end-point of SettingsResource.java to persist the donor and UNICEF sections?

no dont use a script, there is a [settings API ](https://smartregister.atlassian.net/wiki/spaces/Documentation/pages/1765212165/How+to+interact+with+the+OpenSRP+settings+endpoint) that should be used to populate the oneoff donors and unicef sections.

File: src/main/java/org/opensrp/service/StockService.java [hunk 4]
```diff
@@ -185,6 +216,114 @@ public void deleteStock(Long id) {
...
+			logger.error("Exception occurred : " + e.getMessage(), e);
+			failedRecordSummary = new FailedRecordSummary();
+			List<String> validationError = new ArrayList<>();
+			failedRecordSummary.setRowNumber(rowCount);
+			validationError.add("Unknown error occurred");
```
Comment (githengi) [inline src/main/java/org/opensrp/service/StockService.java:236]: Since we know this is a parse exception can we at least reflect that and provide more details

Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/StockService.java:236]: Updating error message to Parse Exception occurred.

**Review by githengi — APPROVED** [review]
